File: app/src/main/java/com/example/kr_kotlin/ui/cart_java/CartViewModel.java
package com.example.kr_kotlin.ui.cart_java;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.kr_kotlin.ui.sub_catalog_java.Product;
import com.example.kr_kotlin.ui.sub_catalog_java.ProductRepository;

import java.util.List;

public class CartViewModel extends ViewModel {

    private final ProductRepository repository = new ProductRepository();

    private final MutableLiveData<List<Product>> _cart = new MutableLiveData<>();
    public LiveData<List<Product>> cart = _cart;

    public void delFromCart(Product product) {
        repository.removeFromCart(product.getId(), this::loadCart);
    }

    public void buyButton(Product product) {
        repository.addToCart(product, this::loadCart);
    }

    public void orderButton() {
        repository.order(_cart.getValue() != null ? _cart.getValue() : List.of(), this::clearCart);
    }

    public void clearCart() {
        List<Product> currentCart = _cart.getValue();
        if (currentCart != null) {
            for (Product product : currentCart) {
                repository.removeFromCart(product.getId(), this::loadCart);
            }
        }
    }

    public void loadCart() {
        repository.getCart(cart -> _cart.setValue(cart));
    }
}

File: app/src/main/java/com/example/kr_kotlin/ui/cart/CartViewModel.kt
package com.example.kr_kotlin.ui.cart

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.map
import com.example.kr_kotlin.ui.sub_catalog.Product
import com.example.kr_kotlin.ui.sub_catalog.ProductRepository

class CartViewModel : ViewModel() {
    private val repository = ProductRepository()

    private val _cart = MutableLiveData<List<Product>>()
    val cart : LiveData<List<Product>> get() = _cart

    fun delFromCart (product: Product) {
        repository.removeFromCart(product.id) {
            loadCart()
        }

    }

    fun buyButton(product: Product) {
        repository.addToCart(product) {
            loadCart()
        }
    }

//    fun orderButton() {
//        val currentCart = _cart.value.orEmpty()
//        if (currentCart.isEmpty()) return
//
//        repository.order(currentCart) {
//            clearCart()
//        }
//    }

    fun orderButton() {
        repository.order(_cart.value.orEmpty()) {
            clearCart()
        }
    }

    fun clearCart() {
        _cart.value?.forEach { repository.removeFromCart(it.id) { loadCart() } } // мб оптимизорвать:
    // вызвать метод loadCart() 1 раз в конце.
    }

    fun loadCart() {
        repository.getCart { _cart.value =  it}
    }
}